File: python3/simpleExampleWithErrors1.py
#!/usr/bin/env python3

import xmlrpc.client

host = "localhost"
port = "8080"

urlEndPoint="http://"+host+":"+port+"/users" # Correct endpoint

# Test with the incorrect end point
wrongURLEndPoint="http://"+host+":"+port+"/notcorrect"

testUser = "alec" # Test with valid data

try:
    proxy = xmlrpc.client.ServerProxy(wrongURLEndPoint)
    result = proxy.userExists(testUser)  # Call method with incorrect end point
    print("Called userExist() on user {}, result:\n{}".format(
        testUser, result))

except xmlrpc.client.Fault as error:
    print("called userExit with wrong end point.\nReturn fault is {}".format(
        error.faultString))

except xmlrpc.client.ProtocolError as error:
    print("""\nA protocol error occurred\nURL: {}\nHTTP/HTTPS headers: {}
Error code: {}\nError message: {}""".format(
        error.url, error.headers, error.errcode, error.errmsg))

except ConnectionError as error:
    print("\nConnection error. Is the server running? {}".format(error))


proxy = xmlrpc.client.ServerProxy(urlEndPoint) # Now use correct end point

# Pass incorrect arguments to remote procedure
try:
    # Call method with incorrect arguments
    result = proxy.userExists(testUser, 21)
    print("\nCalled userExist() on user {}, result {}".format(testUser, result))
except xmlrpc.client.Fault as error:
    print("\ncalled userExit with incorrect args.\nReturn fault is {}".format(
        error.faultString))
except xmlrpc.client.ProtocolError as error:
    print("""\nA protocol error occurred\nURL: {}\nHTTP/HTTPS headers: {}
          "Error code: {}\nError message: {}""".format(
        error.url, error.headers, error.errcode, error.errmsg))

# One valid request, one for non existant data
for testUser in  ["alec","noOne"]:
    try:
        result = proxy.userExists(testUser)
        print("\nCalled userExist() on user {}, result {}".format(
            testUser, result))
        result = proxy.getUserAllDetails(testUser)
        print("""Called getUserAllDetails() on user {},
UUID is {},
active status is {}""".format(
            testUser, result["UUID"], result["activeStatus"]))
    except xmlrpc.client.Fault as error:
        print("\nCalled users API on user {} failed!\nreason {}".format(
            testUser, error.faultString))
    except xmlrpc.client.ProtocolError as error:
        print("""\nA protocol error occurred
URL: {}\nHTTP/HTTPS headers: {}
Error code: {}\nError message: {}""".format(
            error.url, error.headers, error.errcode, error.errmsg))

File: java/src/main/java/SimpleExampleWithErrors1.java
//package com.github.papercutsoftware.howtoxmlrpcclients;
//
// Adapted from example on https://ws.apache.org/xmlrpc/client.html

import org.apache.xmlrpc.client.XmlRpcClient;
import org.apache.xmlrpc.client.XmlRpcClientConfigImpl;
import org.apache.xmlrpc.XmlRpcException;

import java.net.URL;
import java.net.MalformedURLException;


public class SimpleExampleWithErrors1 {

    private final static String host = "localhost";
    private final static String port = "8080";
    private final static String urlEndPoint = "http://"+host+":"+port+"/users";
    private final static String wrongUrlEndPoint = "http://"+host+":"+port+"/notcorrect";
    private final static String testUser = "alec";


    private static XmlRpcClientConfigImpl config;
    private static XmlRpcClient proxy;

    public static void main(String[] args) throws MalformedURLException {

        config = new XmlRpcClientConfigImpl();
        try {
        config.setServerURL(new URL(wrongUrlEndPoint));
        } catch (MalformedURLException ex) {
          System.out.println("Caught MalformedURL Exception " + ex.getCause());
        }

        proxy= new XmlRpcClient();
        proxy.setConfig(config);

        try {
          Object[] params = new Object[]{testUser};
          boolean res = (boolean) proxy.execute("userExists",params);
          System.out.println("Called userExistss for user " + testUser + ": Result is  " + res);
        } catch (XmlRpcException ex) {

          System.out.println("Caught XML RPC Exception " + ex.getCause());
        } catch (Exception ex) {
          System.out.println("Unhandled Exception " + ex.getMessage());
        }

        config.setServerURL(new URL(urlEndPoint));

        proxy.setConfig(config);

        try {
          Object[] params = new Object[]{testUser, new Integer(21)};
          boolean res = (boolean) proxy.execute("userExists",params);
          System.out.println("Called userExistss for user " + testUser + ": Result is  " + res);
        } catch (XmlRpcException ex) {

          System.out.println("Caught XML RPC Exception " + ex.getMessage());
        } catch (Exception ex) {
          System.out.println("Unhandled Exception " + ex.getMessage());
        }


        String[] p = new String[]{"alec", "noOne"};
        for ( String name : p) {

          try {
            Object[] params = new Object[]{name};
            boolean res = (boolean) proxy.execute("userExists",params);
            System.out.println("Called userExistss for user " + name + ": Result is  " + res);
          } catch (XmlRpcException ex) {

            System.out.println("Caught XML RPC Exception " + ex.getCause());
          } catch (Exception ex) {

            System.out.println("Unhandled Exception " + ex.getMessage());
          }

        }
    }
}
